src/fit.jl

```diff
@@ -147,76 +153,72 @@ function eval_metrics!(
     metrics::Vector{String},
 )
+    now_scores = Dict{String,Vector{Float64}}()
 
     if (iter - 1) % estimator.metric_freq == 0
         if estimator.is_training_metric
-            scores = LGBM_BoosterGetEval(estimator.booster, 0)
-            store_scores!(results, estimator, metrics_store_idx, metrics_stored_count, "training", scores, metrics)
-            print_scores(estimator, iter, "training", n_metrics, scores, metrics, verbosity)
+            now_scores["training"] = LGBM_BoosterGetEval(estimator.booster, 0)
+        end
+        for (test_idx, tests_name) in enumerate(tests_names)
+            now_scores[tests_name] = LGBM_BoosterGetEval(estimator.booster, test_idx)
         end
     end
 
-    # Metrics for test sets
-    if (iter - 1) % estimator.metric_freq == 0 || estimator.early_stopping_round > 0
-        for (test_idx, test_name) in enumerate(tests_names)
-            scores = LGBM_BoosterGetEval(estimator.booster, test_idx)
-            # Check if progress should be stored and/or printed
-            if (iter - 1) % estimator.metric_freq == 0
-                store_scores!(results, estimator, metrics_store_idx, metrics_stored_count, test_name, scores, metrics)
-                print_scores(estimator, iter, test_name, n_metrics, scores, metrics, verbosity)
-            end
-
-            # Check if early stopping is called for
-            @inbounds for metric_idx in eachindex(metrics)
-                maximize_score = bigger_is_better[metric_idx] * scores[metric_idx]
-                if maximize_score > best_score[metric_idx, test_idx]
-                    best_score[metric_idx, test_idx] = maximize_score
-                    best_iter[metric_idx, test_idx] = iter
-                elseif iter - best_iter[metric_idx, test_idx] >= estimator.early_stopping_round
-                    # This will shrink it up to the current stored metric
-                    shrinkresults!(results, metrics_store_idx)
-                    log_info(verbosity, "Early stopping at iteration ", iter,
-                             ", the best iteration round is ", best_iter[metric_idx, test_idx], "\n")
-                    return 1
-                end
+    # check early stopping condition
+    if estimator.early_stopping_round > 0
+        for tests_name in tests_names
+            for (metric_idx, metric) in enumerate(metrics)
+                maximize_score = bigger_is_better[metric] * now_scores[tests_name][metric_idx]
+                # All good if maximize_score is better than previous best
+                if maximize_score > best_score[metric][tests_name]
+                    best_score[metric][tests_name] = maximize_score
+                    best_iter[metric][tests_name] = iter
+                    continue
+                # All good if difference between current and best iter is within early_stopping_round
+                elseif (iter - best_iter[metric][tests_name]) < estimator.early_stopping_round
```

Comment:
wouldn't it be just the case that if iter < early_stopping_round we don't bail?

Reply:
I thought so at first, but I think it does not cover all of the scenarios. Using this sample:

Sample run results:
```
iterations, metric (e.g. auc)
1, 0.4
2, 0.5
3, 0.5
4, 0.6
5, 0.7
6, 0.6
7, 0.7
8, 0.4
9, 0.9
10, 0.8
```
- For these test cases:
  - if early_stopping_round <= 0, run should not stop till the end (so get 0.8)
    - Covered (as we first check `early_stopping_round > 0`)
- However for these cases:
  - if early_stopping_round == 2, run should stop at round 5 (rounds 6 and 7 don't show improvement)
    - If we use `iter` only, when `iter == 6`, early_stopping_round will kick in immediately as `iter < early_stopping_round ` (`6 < 2`) and it won't even consider round 7? So in this case it is a coincidence that the results are as expected...
  - if early_stopping_round == 4, run should not stop till the end (continue to see improvement every 4 iterations till the end)
    - Again if we use just `iter`, then this run will stop at round 4, which is what we don't want. Instead, if we update and compare the difference between `iter` and the `best_iter` round to `early_stopping_round`, then that seems correct to me.

Let me know if i'm missing anything obvious! :D

Reply:
Ok yes I can see it follows the logic of the LightGBM callback here: https://github.com/microsoft/LightGBM/blob/37da4ed755897ba16e065eecf18e31d03927e91b/python-package/lightgbm/callback.py#L236

So the way it works is that it needs to see an improvement for at least early_stopping_round iterations to continue training